Fix: Add -arch x86_64 to CXXFLAGS on macOS ARM64 to Resolve Linker Errors in Mixed C++ and D Tests

gulugulubing opened the pull request:
### Problem
On Apple Silicon, the default architecture for C++ compilers is ARM64,
which conflicts with DMD's x86_64 output. This causes linker errors when
building mixed C++ and D projects during `make test`.

### Solution
This change adds `-arch x86_64` to the CXXFLAGS to ensure the C++ compiler
generates x86_64 objects on macOS.

### Testing
- Verified that the build succeeds on macOS with Apple Silicon.
- Confirmed that the generated objects are x86_64.
- All affected tests now pass without linker errors.

gulugulubing committed:
Add -arch x86_64 to CXXFLAGS to fix architecture conflict on macOS when make test

thewilsonator commented on compiler/test/runnable_cxx/abi_tags.d:
```diff
@@ -7,7 +7,8 @@
  * REQUIRED_ARGS: -extern-std=c++11
  * EXTRA_CPP_SOURCES: abi_tags.cpp
- * CXXFLAGS: -std=c++11
+ * CXXFLAGS: -std=c++11 
```
there is trailing whitespace at the end of this line

gulugulubing replied:
Thank you.I've fixed it and pushed the updated commit.

gulugulubing committed:
Fix: Remove trailing whitespace in abi_tags.d

kinke commented:
https://github.com/dlang/dmd/pull/17035 shows that this isn't needed, on an arm64 macOS runner. The reason AFAIU being that both DMD and the testrunner are x86_64 executables, and in that case, the invoked universal `cc` binary defaults to the x86_64 version.

gulugulubing commented:
> #17035 shows that this isn't needed, on an arm64 macOS runner. The reason AFAIU being that both DMD and the testrunner are x86_64 executables, and in that case, the invoked universal `cc`/`c++` binary defaults to the x86_64 version.

Thank you for your feedback! I understand the discussion in the referenced issue is about building and running DMD natively on ARM64 macOS, which is a different scenario from my PR.

My issue occurs during the testing phase after building, specifically when running `make test`. The problem is that the C++ compiler defaults to generating ARM64 code on Apple Silicon, while DMD generates x86_64 code. This causes linker errors when linking mixed C++ and D code. Here's an example of the error: 

ld: warning: ignoring file '/dlang/dmd/compiler/test/test_results/runnable_cxx/d/test6716.cpp.o': found architecture 'arm64', required architecture 'x86_64'
Undefined symbols for architecture x86_64:
  "_main", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: undefined reference to `_main`
       perhaps a library needs to be added with the `-L` flag or `pragma(lib, ...)`
Error: linker exited with status 1

kinke commented:
> I understand the discussion in the referenced issue is about building and running DMD natively on ARM64 macOS, which is a different scenario from my PR.

Yeah the PR is a bit messy (still in draft) - but what's key is that it adds a `macOS 14` job using latest DMD (and another one using LDC). That is a job that builds an x86_64 DMD (not a native arm64 one) on a M1 runner, and then runs all tests as x86_64, no arm64 involved anywhere. And that job works today, without any changes, especially no `-arch x86_64` switches - the 1st commit alone passed that job (https://github.com/dlang/dmd/pull/17035/commits/835d2b98f0fe7e09eae6e6d64e1b73fc859b93c9).

Which `make` are you using? For CI there, we use `brew install make` and then the `/opt/homebrew/opt/make/libexec/gnubin/make binary`. Just in case there's an arch difference across CI and your box for that executable; it's the top-most process when running `make test`. Oh wait, just seen that CI doesn't run `make test`, but runs the tests manually: https://github.com/dlang/dmd/blob/a4cbc08f5bc1a2f7ce3289103198c473671e94c0/ci/run.sh#L131-L132

So I guess it's plausible that `make` gets in the way and causes these issues on your box. Maybe there's a way to put it into x86_64 mode? (I don't have Apple stuff.)

---

What we definitely don't want are hardcoded `-arch x86_64` flags here in these test files. E.g., LDC supports arm64 too and runs these tests, on native M1 runners.

gulugulubing commented:
> > I understand the discussion in the referenced issue is about building and running DMD natively on ARM64 macOS, which is a different scenario from my PR.
> 
> Yeah the PR is a bit messy (still in draft) - but what's key is that it adds a `macOS 14` job using latest DMD (and another one using LDC). That is a job that builds an x86_64 DMD (not a native arm64 one) on a M1 runner, and then runs all tests as x86_64, no arm64 involved anywhere. And that job works today, without any changes, especially no `-arch x86_64` switches - the 1st commit alone passed that job ([835d2b9](https://github.com/dlang/dmd/commit/835d2b98f0fe7e09eae6e6d64e1b73fc859b93c9)).
> 
> Which `make` are you using? For CI there, we use `brew install make` and then the `/opt/homebrew/opt/make/libexec/gnubin/make` binary. Just in case there's an arch difference across CI and your box for that executable; it's the top-most process when running `make test`. Oh wait, just seen that CI doesn't run `make test`, but runs the tests manually:
> 
> https://github.com/dlang/dmd/blob/a4cbc08f5bc1a2f7ce3289103198c473671e94c0/ci/run.sh#L131-L132
> 
> So I guess it's plausible that `make` gets in the way and causes these issues on your box. Maybe there's a way to put it into x86_64 mode? (I don't have Apple stuff.)
> 
> What we definitely don't want are hardcoded `-arch x86_64` flags here in these test files. E.g., LDC supports arm64 too and runs these tests, on native M1 runners.

My make version is :
GNU Make 4.4.1
Built for aarch64-apple-darwin23.4.0

 As a newcomer to contributing, I'm still getting familiar with the D language build process. Modifying the complex scripts or Makefile to enforce arch -x86_64 feels a bit challenging for me at this stage. Therefore, I tried making the change directly in the test files instead. While I understand this approach may seem less elegant, it appeared to be simpler and more controllable for me as a beginner. I appreciate your understanding and guidance!

kinke commented:
No worries, and thx for contributing! :)

You could add some implicit args after this line: https://github.com/dlang/dmd/blob/a4cbc08f5bc1a2f7ce3289103198c473671e94c0/compiler/test/tools/d_do_test.d#L741

E.g.:
```D
// DMD on macOS: force Apple clang to produce x86_64 code (on arm64 too), like DMD
version (DigitalMars) version (OSX)
    testArgs.cxxflags ~= " -arch x86_64";
```

thewilsonator commented:
CI is green now, this good to go?

kinke commented:
Have you read what I wrote? No hardcoded args in these test files.

thewilsonator commented:
> Have you read what I wrote?

Clearly not.

thewilsonator reviewed (changes requested):
per @kinke No hardcoded args in these test files.

benjones commented:
I worked with the author (one of my students) on trying to figure this out and I think we've hit a wall on this and will likely close it soon.  In general, it's tough to tell which compiler is being used for which steps of the build/test process.  Unless you have some good tips on getting some extra debugging info (make VERBOSE=1 helped less than I hoped, for example).  It's a busy time in the semester, hence the lack of updates.

kinke commented:
Oh right, `d_do_test.d` is compiled with the *host* compiler, so `version (DigitalMars)` cannot be used to make this apply to DMD only. But `envData.compiler` can be used: https://github.com/dlang/dmd/blob/001893d99cacfc79079d2b108a6a999d41763d41/compiler/test/tools/d_do_test.d#L211-L232

gulugulubing commented:
It works out in all the tests in runnable_cxx.
Just append  version (OSX) instead of version (DigitalMars) version (OSX) here, like following:

https://github.com/dlang/dmd/blob/001893d99cacfc79079d2b108a6a999d41763d41/compiler/test/tools/d_do_test.d#L741-L741
version (OSX)
    testArgs.cxxflags ~= " -arch x86_64";



But the issue15574.sh in compilable still failed. I guess https://github.com/dlang/dmd/blob/001893d99cacfc79079d2b108a6a999d41763d41/compiler/test/compilable/issue15574.sh#L51-L51, overwrites the cxxflag.

In the PR, I hardcoded -arch x86_64 in the file:
`
if [[ $OS == *"osx"* ]]; then
    ${CC} -m${MODEL} -arch x86_64 -c -o ${C_FILE}${OBJ} $C_FILE
else
    ${CC} -m${MODEL} -c -o ${C_FILE}${OBJ} $C_FILE
fi
`
I don't know whether it's acceptable.

gulugulubing committed:
delete hardcode args in these test files, except dmd/compiler/test/compilable/issue15574.sh

benjones commented on compiler/.DS_Store:
this shouldn't be part of the commit

benjones commented on compiler/.github/workflows/runnable_cxx.yml:
Why the empty file here?

gulugulubing replied:
Sorry, both of them shouldn't be there. I have deleted them

kinke commented on compiler/test/tools/d_do_test.d:
```diff
@@ -739,6 +739,8 @@ bool gatherTestParameters(ref TestArgs testArgs, string input_dir, string input_
 
     findTestParameter(envData, file, "CXXFLAGS", testArgs.cxxflags);
+    version (OSX)
+        testArgs.cxxflags ~= " -arch x86_64";
```
If you add `if (envData.compiler == "dmd")`, then this won't cause regressions when running native arm64 tests (with the other compilers) anymore.

gulugulubing replied:
Thanks, I have added it.

gulugulubing committed:
chore: remove compiler/.DS_Store and complier/.github from PR

gulugulubing committed:
fix: handle DMD compiler separately
- Add check for 'envData.compiler == \"dmd\"'
- Avoid interference with other compilers"

kinke commented on compiler/test/compilable/issue15574.sh:
```diff
@@ -48,7 +48,11 @@ int main() {
+    ${CC} -m${MODEL} -arch x86_64 -c -o ${C_FILE}${OBJ} $C_FILE
+else
+    ${CC} -m${MODEL} -c -o ${C_FILE}${OBJ} $C_FILE
+fi
```
I can live with this hack, will have to remove it for LDC. FYI @ibuclaw.

benjones commented:
Since @kinke is on board with this, can we get this merged?  Thanks!
